dev_setup.sh

```diff
@@ -0,0 +1,155 @@
+        then
+            echo
+            echo "An error occurred; please see $LOG"
+            echo "Here are the last 50 lines:"
+            tail -n 50 $LOG
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $ROOT
+    notice "Updating submodules"
+    logged git submodule update --init
+}
+
+function install_ghcup {
+    if ! ghcup --version &> /dev/null
+    then
+        notice "Installing ghcup, GHC, and cabal"
+        # Technically the installation only requires cabal, but it's
+        # recommended to get the whole GCH shebang in one package.
+        # The output is not routed to log because the installer is interactive.
+        curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh
+    else
+        notice "Using existing ghcup installation"
+    fi
+
+}
+
+function install_gmp {
+    if [ $CRYPTOL_PLATFORM = $MACOS12 ] || [ $CRYPTOL_PLATFORM = $MACOS14 ]; then
+        notice "Installing GMP via Homebrew, if it's not already installed"
+        logged brew install gmp
+
+        # `brew --prefix` is different on macOS 12 and macOS 14
+        echo "export CPATH=$(brew --prefix)/include" >> $VAR_FILE
+        echo "export LIBRARY_PATH=$(brew --prefix)/lib" >> $VAR_FILE
+        notice "You may need to source new environment variables added here:\n" \
+            "\$ source $VAR_FILE"
+    else
+        notice "Did not install GMP. This script only supports macOS 14"
+    fi
+}
+
+# Installs the two solvers required to run the test suite for the repo.
+# Users may want to install other solvers, and indeed the what4 solvers repo
+# includes a half dozen other solvers that are compatible with cryptol. 
+function install_what4_solvers {
+    if ! cvc4 --version &> /dev/null || ! cvc5 --version &> /dev/null; then
+        notice "Installing cvc4 and/or cvc5 solvers"
+
+        if [ $CRYPTOL_PLATFORM = $MACOS14 ]; then
+            solvers_version=$WHAT4_SOLVERS_MACOS_14
+        elif [ $CRYPTOL_PLATFORM = $MACOS12 ]; then
+            solvers_version=$WHAT4_SOLVERS_MACOS_12
+        fi
+
+        solvers_dir=$(mktemp -d)
+        logged curl --proto '=https' --tlsv1.2 -sSfL -o "$solvers_dir/solvers.bin.zip" "$WHAT4_SOLVERS_URL$solvers_version"
+        cd $solvers_dir
+        logged unzip solvers.bin.zip
+        
+        # If we want to install more solvers by default, we can do so here,
+        # although we might need a different check than `--version`
+        for solver in cvc4 cvc5; do
+            if ! $solver --version &> /dev/null; then 
+                notice "Installing $solver"
+                chmod u+x $solver
+                sudo mv $solver /usr/local/bin
+            fi
+        done
+
+        rm -r $solvers_dir
+    else
+        notice "Not installing cvc4 or cvc5 solvers because they already exist"
+
+        if ! (grep -q "version 1.8" <<< "$(cvc4 --version)"); then
+            notice "Your version of cvc4 is incorrect; expected 1.8"
```

Comment:
I'd suggest "unexpected" rather than "incorrect" and adding the path where the corresponding solver was found.  

You might also want to let the user know that the setup is not marked as failure due to these version mismatches but that they might wish to re-run after uninstalling the existing versions  to get the suggested versions installed to avoid runtime cryptol issues.

And lastly, you might want to use vars (e.g. WHAT4_CVC5_VERSION) here that you set near line 30 so that it's close to the declared snapshot version used for future updates.

Reply:
These are good suggestions, I cleaned all this up.